In [1]:
%matplotlib inline  
import pandas as pd 
import numpy as np  
import matplotlib.pyplot as plt
import seaborn as sns
df=pd.read_csv('ANFR_Herault_Mesures.csv')
df.head()

Numéro de mesure  Longitude  Latitude  \
0               175    3.91611  43.58528   
1               180    4.04194  43.64694   
2               543    3.03167  43.36028   
3               545    3.03278  43.36167   
4               850    3.56444  43.64583   

                                    Position  Insee  \
0  03° 54' 58.00'' Est, 43° 35' 07.00'' Nord  34129   
1  04° 02' 31.00'' Est, 43° 38' 49.00'' Nord  34176   
2  03° 01' 54.00'' Est, 43° 21' 37.00'' Nord  34225   
3  03° 01' 58.00'' Est, 43° 21' 42.00'' Nord  34225   
4  03° 33' 52.00'' Est, 43° 38' 45.00'' Nord  34114   

                               Adresse                 Complément  \
0                           Mas de Gau                        NaN   
1  5539 Rue Lou Claou de la Oulivettas                        NaN   
2                 Chemin de Montfaucon  résidence de  Mme SOUTADE   
3                   Chemin de Quarante  résidence de  M. GONZALEZ   
4           480 Chemin de Pioch Courbi      Chez Mme et M. GALOIS   

   Code postal       Commune Date de mesure Mesure réalisée  \
0        34970        LATTES     06/12/2001    En extérieur   
1        34130      MUDAISON     06/12/2001    En extérieur   
2        34620  PUISSERGUIER     23/05/2002    En extérieur   
3        34620  PUISSERGUIER     23/05/2002    En extérieur   
4        34150        GIGNAC     10/10/2002    En extérieur   

              Rapport Conformité de la mesure  Niveau global  \
0  FicheMesure175.pdf                     Oui           0.00   
1  FicheMesure180.pdf                     Oui           0.00   
2  FicheMesure543.pdf                     Oui           0.54   
3  FicheMesure545.pdf                     Oui           0.20   
4  FicheMesure850.pdf                     Oui           0.53   

    Type d'environnement Laboratoire  
0                 Divers     SATELIT  
1                 Divers     SATELIT  
2                 Divers       APAVE  
3                 Divers       APAVE  
4  Rue / Route / Parking       APAVE

In [2]:
df.rename(columns={'Niveau global': 'Niveau_global'}, inplace=True)

In [3]:
df=df.assign(target=1) #création d'une nouvelle colonne
print(df)

     Numéro de mesure  Longitude  Latitude  \
0                 175    3.91611  43.58528   
1                 180    4.04194  43.64694   
2                 543    3.03167  43.36028   
3                 545    3.03278  43.36167   
4                 850    3.56444  43.64583   
..                ...        ...       ...   
907            159667    3.87667  43.59809   
908            160107    3.86363  43.60516   
909            160146    3.22983  43.36141   
910            160274    3.86367  43.60505   
911            160278    3.48046  43.28074   

                                      Position  Insee  \
0    03° 54' 58.00'' Est, 43° 35' 07.00'' Nord  34129   
1    04° 02' 31.00'' Est, 43° 38' 49.00'' Nord  34176   
2    03° 01' 54.00'' Est, 43° 21' 37.00'' Nord  34225   
3    03° 01' 58.00'' Est, 43° 21' 42.00'' Nord  34225   
4    03° 33' 52.00'' Est, 43° 38' 45.00'' Nord  34114   
..                                         ...    ...   
907  03° 52' 36.02'' Est, 43° 35' 53.11'' Nord  

### On a crée une nouvelle colonne "Target" . Par défaut, toute la colonne est égale à 1. Alors, j'ai 912 lignes et 17 colonnes.

In [4]:
df.columns # cette ligne permet d'obtenir les différentes colonnes que j'ai dans Dataset.

Index(['Numéro de mesure', 'Longitude', 'Latitude', 'Position', 'Insee',
       'Adresse', 'Complément', 'Code postal', 'Commune', 'Date de mesure',
       'Mesure réalisée', 'Rapport', 'Conformité de la mesure',
       'Niveau_global', 'Type d'environnement', 'Laboratoire', 'target'],
      dtype='object')

In [5]:
def target():
    for t in range(0,len(df)):
        if df['Niveau_global'][t] > 6 : 
            df['target'][t] = 0
            print(df['Niveau_global'][t]) 

### La fonction target permet de vérifier si le Niveau_global > 6 alors Target=0 

In [6]:
target()

9.63
8.44
8.03
9.06
9.05
6.65
15.66
25.39
10.33
9.1
6.98
21.03
10.57
8.52


<ipython-input-5-43678db2649d>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['target'][t] = 0


In [7]:
pd.crosstab(df.target, "freq") 

col_0   freq
target      
0         14
1        898

### J'obtient :  14 valeurs supérieur à 6 et 898 valeurs inférieurs à 6 . (seuil = 6)

In [8]:
df.head(30) # Les 30 lignes premiers 

Numéro de mesure  Longitude  Latitude  \
0                175    3.91611  43.58528   
1                180    4.04194  43.64694   
2                543    3.03167  43.36028   
3                545    3.03278  43.36167   
4                850    3.56444  43.64583   
5                851    3.56306  43.64806   
6                852    3.56306  43.64694   
7               1072    3.91611  43.58500   
8               1073    3.91667  43.58667   
9               1074    3.91889  43.58583   
10              1115    3.20806  43.34861   
11              1193    3.89750  43.63806   
12              1194    3.89778  43.63889   
13              1195    3.89889  43.63861   
14              1955    3.88556  43.71639   
15              2286    3.86944  43.60833   
16              2428    4.08361  43.55861   
17              2429    4.08361  43.55861   
18              3043    3.23056  43.36139   
19              3045    3.23028  43.36167   
20              3046    3.22861  43.36306   
21              3047    3.22917  43.36222   
22              3261    3.47833  43.30778   
23              3263    3.47750  43.30750   
24              3309    3.79389  43.48944   
25              3310    3.79389  43.48944   
26              3876    3.81139  43.64000   
27              3877    3.81139  43.64000   
28              3878    3.81111  43.64028   
29              3879    3.81167  43.63972   

                                     Position  Insee  \
0   03° 54' 58.00'' Est, 43° 35' 07.00'' Nord  34129   
1   04° 02' 31.00'' Est, 43° 38' 49.00'' Nord  34176   
2   03° 01' 54.00'' Est, 43° 21' 37.00'' Nord  34225   
3   03° 01' 58.00'' Est, 43° 21' 42.00'' Nord  34225   
4   03° 33' 52.00'' Est, 43° 38' 45.00'' Nord  34114   
5   03° 33' 47.00'' Est, 43° 38' 53.00'' Nord  34114   
6   03° 33' 47.00'' Est, 43° 38' 49.00'' Nord  34114   
7   03° 54' 58.00'' Est, 43° 35' 06.00'' Nord  34129   
8   03° 55' 00.00'' Est, 43° 35' 12.00'' Nord  34129   
9   03° 55' 08.00'' Est, 43° 35' 09.00'' Nord  34129   
10  03° 12' 29.00'' Est, 43° 20' 55.00'' Nord  34032   
11  03° 53' 51.00'' Est, 43° 38' 17.00'' Nord  34057   
12  03° 53' 52.00'' Est, 43° 38' 20.00'' Nord  34057   
13  03° 53' 56.00'' Est, 43° 38' 19.00'' Nord  34057   
14  03° 53' 08.00'' Est, 43° 42' 59.00'' Nord  34014   
15  03° 52' 10.00'' Est, 43° 36' 30.00'' Nord  34172   
16  04° 05' 01.00'' Est, 43° 33' 31.00'' Nord  34344   
17  04° 05' 01.00'' Est, 43° 33' 31.00'' Nord  34344   
18  03° 13' 50.00'' Est, 43° 21' 41.00'' Nord  34032   
19  03° 13' 49.00'' Est, 43° 21' 42.00'' Nord  34032   
20  03° 13' 43.00'' Est, 43° 21' 47.00'' Nord  34032   
21  03° 13' 45.00'' Est, 43° 21' 44.00'' Nord  34032   
22  03° 28' 42.00'' Est, 43° 18' 28.00'' Nord  34003   
23  03° 28' 39.00'' Est, 43° 18' 27.00'' Nord  34003   
24  03° 47' 38.00'' Est, 43° 29' 22.00'' Nord  34333   
25  03° 47' 38.00'' Est, 43° 29' 22.00'' Nord  34333   
26  03° 48' 41.00'' Est, 43° 38' 24.00'' Nord  34172   
27  03° 48' 41.00'' Est, 43° 38' 24.00'' Nord  34172   
28  03° 48' 40.00'' Est, 43° 38' 25.00'' Nord  34172   
29  03° 48' 42.00'' Est, 43° 38' 23.00'' Nord  34172   

                                      Adresse  \
0                                  Mas de Gau   
1         5539 Rue Lou Claou de la Oulivettas   
2                        Chemin de Montfaucon   
3                          Chemin de Quarante   
4                  480 Chemin de Pioch Courbi   
5                                Vieux chemin   
6                  401 Chemin de Pioch Courbi   
7                 Château d'eau de Boirargues   
8       Avenue de la Mer - Garage Tunning XXS   
9                       Chemin des Mas de Gau   
10                          15 Rue des Cygnes   
11                256 Chemin du Château d'eau   
12                 50 Chemin du Château d'eau   
13                       Impasse de l'Oppudum   
14                     283 Rue de Montpellier   
15                             Cours Gambetta   
16                            140 Rue 

In [9]:
X=df[['Numéro de mesure','Niveau_global']].to_numpy() # x a 2 colonnes : Numéro de mesure et Niveau_global
print(X)
y=df['target'].to_numpy() # y = est un vecteur contient Target
print(y)

[[1.75000e+02 0.00000e+00]
 [1.80000e+02 0.00000e+00]
 [5.43000e+02 5.40000e-01]
 ...
 [1.60146e+05 3.27000e+00]
 [1.60274e+05 9.20000e-01]
 [1.60278e+05 1.29000e+00]]
[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 

### ==> Ma Dataset n'est pas équilibré car on a plus des 1 que les 0 . On 14 zéros 

### Maintenant , on va faire l'entrainement de mon modéle  

In [10]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0) 

### - On a 30% pour le teste et 70% pour le train 

In [11]:
from sklearn.linear_model import LogisticRegression 
#Create a l.Reg Classifier
logreg1 = LogisticRegression()
#Train the model using the training sets
logreg1.fit(X_train, y_train)
y_pred=logreg1.predict(X_test)

### Le modéle de machine learning que nous allons utiliser est 'LOGISTIC REGRESSION' 

In [12]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
print(confusion_matrix(y_test,y_pred))
print(classification_report(y_test,y_pred))
print(accuracy_score(y_test, y_pred))

[[  3   0]
 [  0 271]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00         3
           1       1.00      1.00      1.00       271

    accuracy                           1.00       274
   macro avg       1.00      1.00      1.00       274
weighted avg       1.00      1.00      1.00       274

1.0


### LOGISTIC REGRESSION peut donner jusqu'à 80% de précision , on peut dire que c'est un bon modéle.

test du modèle avec des nouvelles données

# La nouvelle Dataset 

In [13]:
df1=pd.read_csv('Mesures_nouvelles.csv') # j'ai ajouté une nouvelle Dataset equilibré 
df1.head()

Numéro de mesure  Longitude      Latitude  \
0               175    3.91611      43.58528   
1               180    4.04194      43.64694   
2               543    3.03167      43.36028   
3               545    3.03278      43.36167   
4               850    3.56444      43.64583   

                                    Position  Insee  \
0  03° 54' 58.00'' Est, 43° 35' 07.00'' Nord  34129   
1  04° 02' 31.00'' Est, 43° 38' 49.00'' Nord  34176   
2  03° 01' 54.00'' Est, 43° 21' 37.00'' Nord  34225   
3  03° 01' 58.00'' Est, 43° 21' 42.00'' Nord  34225   
4  03° 33' 52.00'' Est, 43° 38' 45.00'' Nord  34114   

                               Adresse                 Complément  \
0                           Mas de Gau                        NaN   
1  5539 Rue Lou Claou de la Oulivettas                        NaN   
2                 Chemin de Montfaucon  résidence de  Mme SOUTADE   
3                   Chemin de Quarante  résidence de  M. GONZALEZ   
4           480 Chemin de Pioch Courbi      Chez Mme et M. GALOIS   

   Code postal       Commune Date de mesure Mesure réalisée  \
0        34970        LATTES     06/12/2001    En extérieur   
1        34130      MUDAISON     06/12/2001    En extérieur   
2        34620  PUISSERGUIER     23/05/2002    En extérieur   
3        34620  PUISSERGUIER     23/05/2002    En extérieur   
4        34150        GIGNAC     10/10/2002    En extérieur   

              Rapport Conformité de la mesure  target  Niveau global  \
0  FicheMesure175.pdf                     Oui       1           0.23   
1  FicheMesure180.pdf                     Oui       1           0.22   
2  FicheMesure543.pdf                     Oui       1           0.54   
3  FicheMesure545.pdf                     Oui       1           5.55   
4  FicheMesure850.pdf                     Oui       0           6.89   

    Type d'environnement Laboratoire  
0                 Divers     SATELIT  
1                 Divers     SATELIT  
2                 Divers       APAVE  
3                 Divers       APAVE  
4  Rue / Route / Parking       APAVE

### Dans la nouvelle Dataset , j'ai choisie des valeurs proches du seuil , dans la colonne Target pour que le test soit plus fiable et on met en difficulté le modèle.

In [14]:
df1.rename(columns={'Niveau global': 'Niveau_global'}, inplace=True)

In [15]:
Z=df1[['Numéro de mesure','Niveau_global']].to_numpy()
print(Z)
z=df1['target'].to_numpy()
print(z)

[[1.750e+02 2.300e-01]
 [1.800e+02 2.200e-01]
 [5.430e+02 5.400e-01]
 [5.450e+02 5.550e+00]
 [8.500e+02 6.890e+00]
 [8.510e+02 6.900e+00]
 [8.520e+02 5.990e+00]
 [1.072e+03 6.700e+00]
 [1.073e+03 2.200e-01]
 [1.074e+03 6.880e+00]
 [1.115e+03 6.990e+00]
 [1.193e+03 6.350e+00]]
[1 1 1 1 0 0 1 0 1 0 0 0]


In [16]:
y1_pred=logreg1.predict(Z)
print(y1_pred)

[1 1 1 1 0 0 1 0 1 0 0 1]


In [17]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
print(confusion_matrix(z,y1_pred))
print(classification_report(z,y1_pred))
print(accuracy_score(z, y1_pred))

[[5 1]
 [0 6]]
              precision    recall  f1-score   support

           0       1.00      0.83      0.91         6
           1       0.86      1.00      0.92         6

    accuracy                           0.92        12
   macro avg       0.93      0.92      0.92        12
weighted avg       0.93      0.92      0.92        12

0.9166666666666666


### Avec la nouvelle Dataset équilibré , j'obtient une précision de 91% , que c'est une tres bonne résultat.